File: python/spider_browser/protocol/bidi_session.py
"""WebDriver BiDi session over the Spider WebSocket transport (Firefox)."""

from __future__ import annotations

import asyncio
import json
import math
from typing import Any

from .transport import Transport
from ..utils.errors import ProtocolError, TimeoutError
from ..utils.logger import logger

_CMD_TIMEOUT = 30.0


class BiDiSession:
    def __init__(self, transport: Transport, *, command_timeout: float = _CMD_TIMEOUT) -> None:
        self._transport = transport
        self._next_id = 1
        self._pending: dict[int, asyncio.Future[dict[str, Any]]] = {}
        self._event_handlers: dict[str, list] = {}
        self._browsing_context: str | None = None
        self._timeout = command_timeout

    @property
    def context(self) -> str | None:
        return self._browsing_context

    def handle_message(self, data: str) -> bool:
        try:
            msg = json.loads(data)
        except json.JSONDecodeError:
            return False

        if isinstance(msg.get("id"), int) and "type" in msg:
            fut = self._pending.pop(msg["id"], None)
            if fut and not fut.done():
                fut.set_result(msg)
                return True
            return False

        if msg.get("type") == "event" and isinstance(msg.get("method"), str):
            params = msg.get("params", {})
            for h in list(self._event_handlers.get(msg["method"], [])):
                try:
                    h(params)
                except Exception:
                    pass
            return True

        return False

    async def send(self, method: str, params: dict[str, Any]) -> dict[str, Any]:
        cmd_id = self._next_id
        self._next_id += 1
        cmd = {"id": cmd_id, "method": method, "params": params}
        fut: asyncio.Future[dict[str, Any]] = asyncio.get_event_loop().create_future()
        self._pending[cmd_id] = fut
        await self._transport.send_async(json.dumps(cmd))
        try:
            return await asyncio.wait_for(fut, timeout=self._timeout)
        except asyncio.TimeoutError:
            self._pending.pop(cmd_id, None)
            raise TimeoutError(f"BiDi command timeout: {method} ({self._timeout}s)")

    def on(self, method: str, handler: Any) -> None:
        self._event_handlers.setdefault(method, []).append(handler)

    def off(self, method: str, handler: Any) -> None:
        try:
            self._event_handlers.get(method, []).remove(handler)
        except ValueError:
            pass

    # ------------------------------------------------------------------
    # Context management
    # ------------------------------------------------------------------

    async def get_or_create_context(self) -> str:
        if self._browsing_context:
            return self._browsing_context

        # Strategy 1: getTree with short timeout
        try:
            saved = self._timeout
            self._timeout = 5.0
            resp = await self.send("browsingContext.getTree", {})
            self._timeout = saved
            contexts = (resp.get("result") or {}).get("contexts", [])
            if contexts:
                self._browsing_context = contexts[0]["context"]
                return self._browsing_context
        except Exception:
            self._timeout = _CMD_TIMEOUT

        # Strategy 2: create
        try:
            saved = self._timeout
            self._timeout = 5.0
            cr = await self.send("browsingContext.create", {"type": "tab"})
            self._timeout = saved
            ctx = (cr.get("result") or {}).get("context")
            if ctx:
                self._browsing_context = ctx
                return self._browsing_context
        except Exception:
            self._timeout = _CMD_TIMEOUT

        # Strategy 3: placeholder — geckodriver already has one
        self._browsing_context = "__default__"
        return self._browsing_context

    # ------------------------------------------------------------------
    # High-level
    # ------------------------------------------------------------------

    async def navigate(self, url: str) -> None:
        ctx = await self.get_or_create_context()
        await self.send("browsingContext.navigate", {"context": ctx, "url": url, "wait": "complete"})

    async def capture_screenshot(self) -> str:
        ctx = await self.get_or_create_context()
        resp = await self.send("browsingContext.captureScreenshot", {"context": ctx})
        data = (resp.get("result") or {}).get("data")
        if not isinstance(data, str):
            raise ProtocolError("captureScreenshot: missing result.data")
        return data

    async def evaluate(self, expression: str) -> Any:
        ctx = await self.get_or_create_context()
        resp = await self.send("script.evaluate", {
            "expression": expression,
            "target": {"context": ctx},
            "awaitPromise": False,
            "resultOwnership": "none",
        })
        if resp.get("type") == "error":
            raise ProtocolError(f"BiDi script error: {resp.get('message') or resp.get('error')}")
        result_obj = (resp.get("result") or {}).get("result") or resp.get("result")
        return self._extract_bidi_value(result_obj)

    async def get_html(self) -> str:
        return await self.evaluate("document.documentElement.outerHTML")  # type: ignore

    async def perform_actions(self, actions: list[dict[str, Any]]) -> None:
        ctx = await self.get_or_create_context()
        await self.send("input.performActions", {"context": ctx, "actions": actions})

    async def click_point(self, x: float, y: float) -> None:
        await self.perform_actions([{
            "type": "pointer", "id": "mouse",
            "actions": [
                {"type": "pointerMove", "x": round(x), "y": round(y)},
                {"type": "pointerDown", "button": 0},
                {"type": "pointerUp", "button": 0},
            ],
        }])

    async def insert_text(self, text: str) -> None:
        actions = []
        for ch in text:
            actions.extend([{"type": "keyDown", "value": ch}, {"type": "keyUp", "value": ch}])
        await self.perform_actions([{"type": "key", "id": "keyboard", "actions": actions}])

    def destroy(self) -> None:
        for fut in self._pending.values():
            if not fut.done():
                fut.set_exception(Exception("session destroyed"))
        self._pending.clear()
        self._event_handlers.clear()

    # ------------------------------------------------------------------

    def _extract_bidi_value(self, rv: Any) -> Any:
        if not rv or not isinstance(rv, dict):
            return rv
        t = rv.get("type")
        if t in ("undefined",):
            return None
        if t == "null":
            return None
        if t in ("string", "number", "boolean", "bigint"):
            return rv.get("value")
        if t == "array" and isinstance(rv.get("value"), list):
            return [self._extract_bidi_value(v) for v in rv["value"]]
        if t == "object" and isinstance(rv.get("value"), list):
            obj: dict[str, Any] = {}
            for entry in rv["value"]:
                if isinstance(entry, list) and len(entry) == 2:
                    k = entry[0] if isinstance(entry[0], str) else (entry[0] or {}).get("value", str(entry[0]))
                    obj[k] = self._extract_bidi_value(entry[1])
            return obj
        return rv.get("value", rv)

File: go/bidi_session.go
package spiderbrowser

import (
	"encoding/json"
	"fmt"
	"sync"
	"sync/atomic"
	"time"
)

type bidiPendingRequest struct {
	ch    chan *BiDiResponse
	timer *time.Timer
}

// BiDiSession manages a WebDriver BiDi session (for Firefox).
type BiDiSession struct {
	mu             sync.Mutex
	nextID         int64
	pending        sync.Map // int -> *bidiPendingRequest
	eventHandlers  sync.Map // string -> []func(map[string]any)
	transport      *Transport
	contextID      string
	commandTimeout time.Duration
}

// NewBiDiSession creates a new BiDi session.
func NewBiDiSession(transport *Transport, commandTimeout time.Duration) *BiDiSession {
	if commandTimeout == 0 {
		commandTimeout = 30 * time.Second
	}
	return &BiDiSession{
		transport:      transport,
		commandTimeout: commandTimeout,
	}
}

// HandleMessage processes a raw message from the transport.
func (s *BiDiSession) HandleMessage(data string) bool {
	var raw map[string]any
	if err := json.Unmarshal([]byte(data), &raw); err != nil {
		return false
	}

	// Response (has "id")
	if idRaw, ok := raw["id"]; ok {
		id := toInt(idRaw)
		if val, loaded := s.pending.LoadAndDelete(id); loaded {
			pr := val.(*bidiPendingRequest)
			pr.timer.Stop()
			var resp BiDiResponse
			json.Unmarshal([]byte(data), &resp)
			pr.ch <- &resp
			return true
		}
		return false
	}

	// Event (type === "event")
	if tp, ok := raw["type"].(string); ok && tp == "event" {
		method, _ := raw["method"].(string)
		params := map[string]any{}
		if p, ok := raw["params"].(map[string]any); ok {
			params = p
		}
		s.fireEvent(method, params)
		return true
	}

	return false
}

// Send sends a BiDi command and waits for the response.
func (s *BiDiSession) Send(method string, params map[string]any) (*BiDiResponse, error) {
	id := int(atomic.AddInt64(&s.nextID, 1))
	if params == nil {
		params = map[string]any{}
	}
	cmd := BiDiCommand{
		ID:     id,
		Method: method,
		Params: params,
	}

	cmdJSON, err := json.Marshal(cmd)
	if err != nil {
		return nil, err
	}

	pr := &bidiPendingRequest{
		ch: make(chan *BiDiResponse, 1),
	}
	pr.timer = time.AfterFunc(s.commandTimeout, func() {
		if _, loaded := s.pending.LoadAndDelete(id); loaded {
			pr.ch <- nil
		}
	})

	s.pending.Store(id, pr)
	if err := s.transport.Send(string(cmdJSON)); err != nil {
		s.pending.Delete(id)
		pr.timer.Stop()
		return nil, err
	}

	resp := <-pr.ch
	if resp == nil {
		return nil, newTimeoutError(fmt.Sprintf("BiDi command timeout: %s (%v)", method, s.commandTimeout))
	}
	if resp.Type == "error" {
		return nil, newProtocolError(fmt.Sprintf("BiDi error: %s: %s", resp.Error, resp.Message))
	}
	return resp, nil
}

// On subscribes to a BiDi event.
func (s *BiDiSession) On(method string, handler func(map[string]any)) {
	val, _ := s.eventHandlers.LoadOrStore(method, &[]func(map[string]any){})
	handlers := val.(*[]func(map[string]any))
	s.mu.Lock()
	*handlers = append(*handlers, handler)
	s.mu.Unlock()
}

// Off removes a BiDi event handler.
func (s *BiDiSession) Off(method string, handler func(map[string]any)) {
	val, ok := s.eventHandlers.Load(method)
	if !ok {
		return
	}
	handlers := val.(*[]func(map[string]any))
	s.mu.Lock()
	if len(*handlers) > 0 {
		*handlers = (*handlers)[:len(*handlers)-1]
	}
	s.mu.Unlock()
}

func (s *BiDiSession) fireEvent(method string, params map[string]any) {
	val, ok := s.eventHandlers.Load(method)
	if !ok {
		return
	}
	handlers := val.(*[]func(map[string]any))
	s.mu.Lock()
	hs := make([]func(map[string]any), len(*handlers))
	copy(hs, *handlers)
	s.mu.Unlock()
	for _, h := range hs {
		func() {
			defer func() { recover() }()
			h(params)
		}()
	}
}

// GetOrCreateContext gets or creates a browsing context.
func (s *BiDiSession) GetOrCreateContext() (string, error) {
	// Try to get existing contexts
	resp, err := s.Send("browsingContext.getTree", map[string]any{})
	if err == nil && resp.Result != nil {
		if contexts, ok := resp.Result["contexts"].([]any); ok && len(contexts) > 0 {
			if ctx, ok := contexts[0].(map[string]any); ok {
				if ctxID, ok := ctx["context"].(string); ok {
					s.contextID = ctxID
					return ctxID, nil
				}
			}
		}
	}

	// Create a new context
	resp, err = s.Send("browsingContext.create", map[string]any{
		"type": "tab",
	})
	if err != nil {
		return "", err
	}
	if resp.Result != nil {
		if ctxID, ok := resp.Result["context"].(string); ok {
			s.contextID = ctxID
			return ctxID, nil
		}
	}

	return "", newProtocolError("Failed to create browsing context")
}

// Navigate navigates to a URL in the BiDi context.
func (s *BiDiSession) Navigate(url string) error {
	_, err := s.Send("browsingContext.navigate", map[string]any{
		"context": s.contextID,
		"url":     url,
		"wait":    "complete",
	})
	return err
}

// GetHTML returns the full page HTML via script.evaluate.
func (s *BiDiSession) GetHTML() (string, error) {
	val, err := s.Evaluate("document.documentElement.outerHTML")
	if err != nil {
		return "", err
	}
	if str, ok := val.(string); ok {
		return str, nil
	}
	return "", nil
}

// Evaluate evaluates a JavaScript expression.
func (s *BiDiSession) Evaluate(expression string) (any, error) {
	resp, err := s.Send("script.evaluate", map[string]any{
		"expression":  expression,
		"target":      map[string]any{"context": s.contextID},
		"awaitPromise": true,
		"resultOwnership": "none",
	})
	if err != nil {
		return nil, err
	}
	if resp.Result != nil {
		if result, ok := resp.Result["result"].(map[string]any); ok {
			return extractBiDiValue(result), nil
		}
	}
	return nil, nil
}

// CaptureScreenshot captures a screenshot via BiDi.
func (s *BiDiSession) CaptureScreenshot() (string, error) {
	resp, err := s.Send("browsingContext.captureScreenshot", map[string]any{
		"context": s.contextID,
	})
	if err != nil {
		return "", err
	}
	if resp.Result != nil {
		if data, ok := resp.Result["data"].(string); ok {
			return data, nil
		}
	}
	return "", newProtocolError("captureScreenshot: missing result.data")
}

// ClickPoint clicks at viewport coordinates via BiDi input actions.
func (s *BiDiSession) ClickPoint(x, y float64) error {
	return s.PerformActions([]map[string]any{
		{
			"type": "pointer",
			"id":   "mouse",
			"actions": []map[string]any{
				{"type": "pointerMove", "x": int(x), "y": int(y)},
				{"type": "pointerDown", "button": 0},
				{"type": "pointerUp", "button": 0},
			},
		},
	})
}

// PerformActions performs input actions via BiDi.
func (s *BiDiSession) PerformActions(actions []map[string]any) error {
	_, err := s.Send("input.performActions", map[string]any{
		"context": s.contextID,
		"actions": actions,
	})
	return err
}

// InsertText inserts text via BiDi key actions.
func (s *BiDiSession) InsertText(text string) error {
	actions := make([]map[string]any, 0, len(text)*2)
	for _, ch := range text {
		actions = append(actions,
			map[string]any{"type": "keyDown", "value": string(ch)},
			map[string]any{"type": "keyUp", "value": string(ch)},
		)
	}
	return s.PerformActions([]map[string]any{
		{
			"type":    "key",
			"id":      "keyboard",
			"actions": actions,
		},
	})
}

// Destroy cleans up the BiDi session.
func (s *BiDiSession) Destroy() {
	s.pending.Range(func(key, value any) bool {
		pr := value.(*bidiPendingRequest)
		pr.timer.Stop()
		close(pr.ch)
		s.pending.Delete(key)
		return true
	})
	s.eventHandlers = sync.Map{}
}

// extractBiDiValue extracts a value from a BiDi result object.
func extractBiDiValue(result map[string]any) any {
	tp, _ := result["type"].(string)
	switch tp {
	case "string":
		return result["value"]
	case "number":
		return result["value"]
	case "boolean":
		return result["value"]
	case "null", "undefined":
		return nil
	case "array":
		if arr, ok := result["value"].([]any); ok {
			out := make([]any, len(arr))
			for i, v := range arr {
				if m, ok := v.(map[string]any); ok {
					out[i] = extractBiDiValue(m)
				} else {
					out[i] = v
				}
			}
			return out
		}
	case "object":
		if pairs, ok := result["value"].([]any); ok {
			out := make(map[string]any)
			for _, pair := range pairs {
				if p, ok := pair.([]any); ok && len(p) == 2 {
					key, _ := p[0].(string)
					if vm, ok := p[1].(map[string]any); ok {
						out[key] = extractBiDiValue(vm)
					} else {
						out[key] = p[1]
					}
				}
			}
			return out
		}
	}
	return result["value"]
}
